### ISSUE TITLE
Extend the ConnectionPool Example to showcase using the transactor in a larger app

### ISSUE THREAD

**Issue by bmnn-mchl** [tpolecat/doobie#795]
Thank you very much for creating doobie!
Hope you don't mind me posting such a noob question, but I am just not sure if what I intend to do is a dump idea or the way to go?
The connection pool [example](https://github.com/tpolecat/doobie/blob/series/0.6.x/modules/example/src/main/scala/example/HikariExample.scala) shows how to create an connection as a ressource

`object HikariApp extends IOApp {
  // Resource yielding a transactor configured with a bounded connect EC and an unbounded
  // transaction EC. Everything will be closed and shut down cleanly after use.
  val transactor: Resource[IO, HikariTransactor[IO]] =
  for {
    ce <- ExecutionContexts.fixedThreadPool[IO](32) // our connect EC
    te <- ExecutionContexts.cachedThreadPool[IO]    // our transaction EC
    xa <- HikariTransactor.newHikariTransactor[IO](
      "org.h2.Driver",                        // driver classname
      "jdbc:h2:mem:test;DB_CLOSE_DELAY=-1",   // connect URL
      "sa",                                   // username
      "",                                     // password
      ce,                                     // await connection here
      te                                      // execute JDBC operations here
    )
  } yield xa

  def run(args: List[String]): IO[ExitCode] =
    transactor.use { xa: HikariTransactor[IO] =>
      // Construct and run your server here!
        sql"""
        create table if not exists numbers(id serial auto_increment primary key,
        value integer not null)
        """.update.run.transact(xa).unsafeRunSync()
      sql"insert into numbers (value) values(1)".update.run.transact(xa).unsafeRunSync()
// possibly pass xa to other functions which need to access the database either synchronously or asynchronously like f(xa, ...), g(xa, ...)
      println(sql"select value from numbers".query[Int].to[List].transact(xa).unsafeRunSync())
     ExitCode.Success.pure[IO]
    }
}` 

Is it correct to pass a reference to the transactor around like this? Are their any caveats that apply?
Many thanks again, any help would be greatly appreciated!

**Comment by Jetbo** [comment]
I ran into this week and found that you should only call `transactor.use { xa => ... }` once and pass around `xa` to the various services that need to make database calls. Otherwise if you call `.use` more than once, it'll start more than on connection pool.

**Comment by droyop** [comment]
Hi! I read your post some days ago and I promised myself I'd come back here at the end of my struggle and try to help.

Happy to accept suggestions. I'm fairly junior in this area.

In my (finally working) example I use the same IOApp pattern. I load the config into a config case class, and migrate the database (Though that's quite irrelevant for the matter)

```scala
object Server extends IOApp {

  //Load the configuration objects from the application.conf file
  implicit private val conf: Config = Config.load().unsafeRunSync()

  //Migrate the database
  Flyway.configure.dataSource(conf.db.url, conf.db.user, conf.db.password).load.migrate
```
... and then I create the CP just like you did

```scala
  //Create the HikariCP as a resource
  implicit private val transactor: Resource[IO, HikariTransactor[IO]] = for {
    ce <- ExecutionContexts.fixedThreadPool[IO](conf.db.threadSize)
    te <- ExecutionContexts.cachedThreadPool[IO]
    xa <- HikariTransactor.newHikariTransactor[IO](conf.db.driver,
      conf.db.url,
      conf.db.user,
      conf.db.password,
      ce,
      te)
  } yield xa
```
... The critical part. We use the resource, and define the transactor as implicit to be used by the repositories, wrapping _everything_.

```scala

 //Run the IOApp
  override def run(args: List[String]): IO[ExitCode] = {
    //We wrap everything with the resource and use the underlying transactor
    //implicitly
    transactor.use{implicit xa =>

      //Create an implicit instance of the repository (taking the implicit transactor)
      implicit val clientRepo: ClientRepository = new ClientRepository()

      //Create the Http4s server
      BlazeServerBuilder[IO]
          //Bind to the host:port
          .bindHttp(conf.srv.port, conf.srv.host)
          //The HttpApp defines the routes & behaviour (We can bind several)
          .withHttpApp(new ClientService().routes)
          .serve
          .compile
          .drain
          .as(ExitCode.Success)
    }
  }
``` 
The repository is fairly simple. It just takes the implicit transactor, and exposes a getClient method which return either the client or a not found error

```scala
class ClientRepository(implicit xa: HikariTransactor[IO]) {

  def getClient(id: Long): IO[Either[ClientNotFoundError.type, Client]] = {
    sql"SELECT id, name FROM client WHERE id = $id"
      .query[Client]
      .option
      .transact(xa)
      .map {
        case Some(client) => Right(client)
        case None => Left(ClientNotFoundError)
      }
  }
```

The service takes the implicit repository, and defines the routes, the logic and response handling. 

In this case it just implements a get method for a client id. It gets the result from the repository, and maps it to either json or a 404 error.

```scala
class ClientService(implicit repository: ClientRepository) extends Http4sDsl[IO] {

  val routes = HttpRoutes.of[IO] {
    case GET -> Root / "clients" / LongVar(id) =>
      for {
        result <- repository.getClient(id)
        response <- buildResponse(result)
      } yield response
  }.orNotFound

  private def buildResponse(result: Either[ClientNotFoundError.type, Client]) = {
    result match {
      case Left(ClientNotFoundError) => NotFound()
      case Right(client) => Ok(client.asJson)
    }
  }
}
```

(just for completeness)
```scala
  case class Client(id: Option[Long], name: String)

  case object ClientNotFoundError
```

I have not tested this properly but based on the debug log everything seems to work really nice with the CP, and the server works like a charm and it's ridiculously lightweight.

Kudos: @jaspervz for the inspiration: https://github.com/jaspervz/todo-http4s-doobie You are the real MVP (I'll try to push the changes with the HikariCP for the newest doobie version)

**Comment by wb14123** [comment]
I also suspect the life cycle of `transactor` is the whole application, otherwise every time we `use` it it will also generate multiple threads.

But what's the point to make it as a `Resource` if the life cycle of it is the whole application?

**Comment by jatcwang** [comment]
Yes that's right. You create the transactor once and use it for your whole application. When you application exits (e.g. ctrl-c) then the connection pool will be released cleanly.

See https://github.com/softwaremill/bootzooka/blob/master/backend/src/main/scala/com/softwaremill/bootzooka/Main.scala for example

**Comment by wonk1132** [comment]
@droyop perhaps you could PR a doc change with your example in it? The docs are pretty misleading on this point.
